design/ceph/osd-migration.md

```diff
@@ -0,0 +1,188 @@
+
+#### OSD Prepare
+
+- The `ceph-volume prepare` command is updated to use backend store provided in the `spec.storage.store`. If no value is provided for `spec.storage.store` then default to `bluestore` backend.
+
+    - RAW MODE
+        ```
+        ceph-volume raw prepare <OSD_STORE_FLAG> --data /dev/vda
+        ```
+
+    - LVM MODE
+      ```
+      ceph-volume lvm prepare <OSD_STORE_FLAG> --data /dev/vda
+      ```
+
+#### OSD Create
+
+- `OSD_STORE_FLAG` should use the value provided in the `spec.storage.store`. If no values are provided by the user, then `OSD_STORE_FLAG` defaults to `bluestore`
+- `ceph-volume` commands use the `OSD_STORE_FLAG` while activating the OSDs.
+
+    - RAW MODE
+      ```
+      ceph-volume raw activate --device "$DEVICE" --no-systemd --no-tmpfs
+      ```
+
+    - LVM MODE
+      ```
+      ceph-volume lvm activate --no-systemd "$OSD_STORE_FLAG" "$OSD_ID" "$OSD_UUID"
+      ```
+
+**NOTE**: ceph-volume `activate` command doesn't require `OSD_STORE_FLAG` to be passed as an argument. It auto detects the backend store that was used during the OSD `prepare`
+
+### Existing OSDs
+
+- After upgrading an existing rook ceph cluster to a Ceph version that supports `bluestore-rdr`, users can upgrade the `bluestore` based OSDs to use `bluestore-rdr`
+- Backend store of the OSDs can not be overridden, so this update will require the OSDs to be replaced one by one.
+- In order to start this osd migration, users need to patch the ceph cluster spec like below:
+
+```yaml
+storage:
+  store:
+    type: bluestore-rdr
+    updateStore: yes-really-update-store
+```
+
+#### Replace OSD
+- Operator reconciler will replace one OSD at a time.
+- A config map will be used to store the list of OSDs to be replaced and the OSD being replaced currently.
+- OSD replacement steps:
+  1. List out all OSDs where `osd-store:<osd store type>` does not match with `spec.storage.store.type`.
+  2. These OSDs will not be upgraded before migrating to the new OSD store. This is required because any upgrade operation will update the osd store to `bluestore-rdr` in the OSD deployment spec without migrating the OSD. As a result, the operator will lose track of the OSDs that need migration.
+  3. If the PG status is not `active+clean`, don't replace any OSD.
+  4. If the PG status is `active+clean` but a previous OSD replacement is not completed, then don't replace any new OSD.
+  5. If the PG status is `active+clean` and no replacement is in progress, then select an OSD to be replaced from the list.
+  6. Delete the OSD deployment.
+  7. Create an OSD prepare job with an environment variable of OSD ID to be replaced.
+  8. OSD prepare pod will destroy the OSD and prepare it again using the same OSD ID. Refer `Destroy OSD` section below for details.
+  9. Once the destroyed OSD pod is added back, delete the configmap.
```

Comment:
If we're deleting the configmap, then we're saying that the OSDInProgress is the only thing we need to persist, right?

Reply:
yes. We are getting the list of OSDs to replace on every reconcile. So `OSDInProgress` is the only think we need to persist.